llmcam/chat_ui.py

```diff
@@ -3,45 +3,221 @@
+import inspect
+
+from typing import Optional, Union
+from types import NoneType
+from .ytlive import capture_youtube_live_frame
+from .gpt4v import ask_gpt4v
+
+# %% ../nbs/05_chat_ui.ipynb 6
+def extract_live_info(
+        link: Optional[str] = None  # YouTube Live link
+    ) -> dict:  # The extracted information
+    """Extract information from a YouTube Live"""
+    if link:
+        image = capture_youtube_live_frame(link)
+    else:
+        image = capture_youtube_live_frame()
+    
+    return ask_gpt4v(image)
+
+# %% ../nbs/05_chat_ui.ipynb 8
+# Extract parameter comments from the function
+def extract_parameter_comments(func):
+    """Extract comments for function arguments"""
+    # Get the source code of the function
+    source = inspect.getsource(func)
+    # Parse the source code into an AST
+    tree = ast.parse(source)
+    
+    # Extract comments for function arguments
+    comments = {}
+    for node in ast.walk(tree):
+        if isinstance(node, ast.FunctionDef) and node.name == func.__name__:
+            # Get arguments and comments in the function
+            for arg in node.args.args:
+                arg_name = arg.arg
+                # Check if there's an inline comment associated with the argument
+                if arg.end_lineno and arg.col_offset:
+                    # Loop through the source code lines to find the comment
+                    lines = source.splitlines()
+                    for line in lines:
+                        if line.strip().startswith(f"{arg_name}:") and "#" in line:
+                            comment = line.split("#")[1].strip()
+                            comments[arg_name] = comment
+    return comments
+
+# %% ../nbs/05_chat_ui.ipynb 13
+def param_converter(
+        param_type,  # The type of the parameter
+        description  # The description of the parameter
+    ) -> dict:  # The converted parameter
+    """Convert parameter types to acceptable types for tool schema"""
+    simple_types = {
+        str: "string",
+        int: "number",
+        float: "number",
+        bool: "boolean",
+    }
+    if param_type in simple_types:
+        return { "type": simple_types[param_type], "description": description }
+    elif param_type == NoneType:
+        return { "type": "null", "description": "A default value will be automatically used." }
+    
+    if hasattr(param_type, '__origin__') and param_type.__origin__ == Union:
+        # Recursively convert the types
+        descriptions = description.split(" or ")
+        subtypes = param_type.__args__
+        if len(subtypes) > len(descriptions):
+            descriptions = descriptions + ["A description is not provided"] * (len(subtypes) - len(descriptions))
+
+        return {
+            "anyOf": [param_converter(subtype, desc) for subtype, desc in zip(subtypes, descriptions)]
+        }
+    return { "type": "string", "description": description }
+
+# %% ../nbs/05_chat_ui.ipynb 17
+def tool_schema(func):
```

Comment:
please modify to import this kind of "function_to_fc" (called, "class_to_fc") element functions "06_class_to_fc.ipynb". It's better to keep chat_ui isolated from any non_ui functions. Also you may want to split into non-GUI chat module and UI part as well.

1. non-GUI chat module should maintain "messages".
2. UI module should only maitain the code to present "messages" and some "button".